Updating overview, added section about Ceci pipelines and stages. (#74)
LSSTDESC/RAIL: docs/source/overview.rst

Overview
********

RAIL enables production of photo-z data products at-scale for LSST as well as 
stress-testing of multiple estimation approaches in the presence of realistically 
complex systematic imperfections in the input photometry and prior information 
(such as template libraries and training sets) under science-agnostic and 
science-specific metrics. By providing both functionalities in the same package, 
the exact same estimation procedure validated through controlled experimentation 
may be applied to real data without loss of validity. To support such an ambitious 
goal, RAIL has a highly modular structure encompassing three aspects of this kind 

> of experiment, enabled by specialized object types, however, the result is that 

Reviewer: ```suggestion
of experiment, enabled by specialized object types; however, the result is that 
```

RAIL is unavoidably complicated. This overview seeks to present the foundational

> structures that underlie RAIL's structure, the overarching organizational 

Reviewer: ```suggestion
building blocks that underlie RAIL's structure, the overarching organizational 
```
(or whichever synonym to avoid the repetition)

philosophy, and the specific types of included functionality.

Introduction to stages and pipelines
************************************

> While all of RAIL's functionality is accessible through jupyter notebooks to 

Reviewer: ```suggestion
While all of RAIL's functionality is accessible through Jupyter notebooks to 
```

facilitate experimentation, RAIL's utility is in being able to run the code
developed and validated under controlled conditions on real data at-scale by
executing scripts on high-performance computing clusters (HPCs).
The tool that RAIL uses to organize these scripts is 
`ceci <https://ceci.readthedocs.io/en/latest/>`_, a workflow management 
package specifically designed for running DESC analyses on HPCs, e.g. using 
`TXPipe <https://github.com/LSSTDESC/TXPipe/>`_. At a very high level, a 

> workflow is a pipelines comprised of stages.

Reviewer: ```suggestion
workflow is a pipeline comprised of stages.
```

**Stages**:
A stage performs one unit of work, defined by its input(s) and output(s), its name, 

> and its stage-specific configuration parameters, that can be parallelized across 

Reviewer: ```suggestion
and its stage-specific configuration parameters. It can be parallelized across 
```
This is probably the most subjective suggestion I'll make, so feel free to ignore it, but I feel that the commas in this sentence reduce its readability. Maybe the following could keep it all one sentence but provide clarity to the antecedent of "that can be"?

A stage performs one unit of work (defined by its input(s) and output(s), its name, 
and its stage-specific configuration parameters) that can be parallelized across ...

earlier stage in the pipeline, followed by a `run()` step to actually perform the 
specified calculations.

Core data structures
********************

**TODO: DataHandle/DataStore should be explained here**

**TODO: essential tables_io functionality should be introduced here**

**`qp.Ensemble` objects**:
Redshift data products may take many forms; probability density functions (PDFs) 
characterizing the redshift distribution of a sample of galaxies or each galaxy 
individually are defined by values of parameters under a choice of 
parameterization. To enable parameterization-agnostic downstream analyses,
the `qp <https://github.com/LSSTDESC/qp>`_ package provides a shared interface 
to many parameterizations of univariate PDFs and utilities for performing 
conversions, evaluating metrics, and executing at-scale input-output operations. 
RAIL stages provide and/or ingest their photo-z data products as ``qp.Ensemble`` 
objects, both for collections of individual galaxies and for the summarized 
redshift distribution of samples of galaxies (such as members of a tomographic 
bin or galaxy cluster members). The key features of a `qp.Ensemble` are the 

> `metadata` of the type of parameteriztion and defining parameters shared by the 

Reviewer: ```suggestion
`metadata` of the type of parameterization and defining parameters shared by the 
```

For a working example illustrating all three components of RAIL, see the 
`examples/goldenspike_examples/goldenspike.ipynb <https://github.com/LSSTDESC/RAIL/blob/main/examples/goldenspike_examples/goldenspike.ipynb>`_ 
Jupyter notebook.

`creation`
==========

The creation subpackage has two main components enabling forward-modeling of 
realistically complex mock data.
The creation modules provide a joint probability space of redshift and photometry, 
and the degradation modules introduce systematic imperfections into galaxy catalogs, 
which can be used to stress-test photo-z estimators. 

**Creation modules**: 
This code enables the generation of mock photometry corresponding to a fully 
self-consistent forward model of the joint probability space of redshift and photometry. 
Beyond simply drawing samples of redshift and photometry defining galaxy catalogs, 
this forward model-based approach can provide a true posterior and likelihood for 
each galaxy, enabling novel metrics for individual galaxies that are not available 
from traditionally simulated catalogs without a notion of inherent uncertainty.

**Creation base design**: 

> To ensure the mutual consistency of RAIL's mock data with known physics but leaving 

Reviewer: ```suggestion
To ensure the mutual consistency of RAIL's mock data with known physics while leaving 
```

open the possibility of surprises beyond current data, we make an interpolative 
and extrapolative model beginning from input data of galaxy redshifts and photometry, 
for example, the DC2 extragalactic catalog.
While multiple generating engines are possible and desired, our initial implementation 
utilizes a normalizing flow via the `pzflow package <https://github.com/jfcrenshaw/pzflow>`_. 
The normalizing flow model fits the joint distribution of redshift and photometry 
(and any other parameters that are supplied as inputs), and galaxy redshifts and 
photometry drawn from that joint probability density will have a true likelihood 
and a true posterior.

**Degradation modules**: 
Degraders build upon the creator models of the probability space of redshift and 
photometry to emulate realistically complex imperfections, such as physical systematics, 
into mock data, which can be used to generate self-consistent photometric training/test 
set pairs.
The high-dimensional probability density outlined in the `creation` directory can 
be modified to reproduce the realistic mismatches between training and test sets, 

> for example inclusion of photometric errors due to observing effects, spectroscopic 

Reviewer: ```suggestion
for example, inclusion of photometric errors due to observing effects, spectroscopic 
```

incompleteness from specific surveys, incorrect assignment of spectroscopic redshift 
due to line confusion, the effects of blending, etc.
Training and test set data may be drawn from such probability spaces with systematics 
applied in isolation, which preserves the existence of true likelihoods and posteriors, 
though applying multiple degraders in series enables more complex selections to 
be built up. 

**Degradation base design**: 
The base design for degraders in our current scheme is that degraders take in a 

> DataFrame (or a creator that can generate samples on the fly) and returns a modified 

Reviewer: ```suggestion
DataFrame (or a creator that can generate samples on the fly) and return a modified 
```
